R/appMetadata.R

```diff
@@ -59,6 +60,53 @@ appMetadata <- function(appDir,
   )
 }
 
+checkAppLayout <- function(appDir, appPrimaryDoc = NULL) {
+  appFilesBase <- tolower(list.files(appDir))
+  wwwFiles <- tolower(list.files(file.path(appDir, "www/")))
+
+  primaryIsRScript <- identical(tolower(tools::file_ext(appPrimaryDoc)), "r")
+
+  # check for single-file app collision
+  if (primaryIsRScript && "app.r" %in% appFilesBase) {
+    cli::cli_abort(
+      "Project must not contain both {.file app.R} and a single-file Shiny app."
+    )
+  }
+
+  # Do some checks for a valid application structure
+  satisfiedLayouts <- c(
+    shinyAndUi = all(c("server.r", "ui.r") %in% appFilesBase),
+    shinyAndIndex = "server.r" %in% appFilesBase && "index.html" %in% wwwFiles,
+    app = primaryIsRScript || any("app.r" %in% appFilesBase),
+    Rmd = any(grepl(glob2rx("*.rmd"), appFilesBase)),
+    Qmd = any(grepl(glob2rx("*.qmd"), appFilesBase)),
+    static = any(grepl("(?:html?|pdf)$", appFilesBase)),
+    plumber = any(c("entrypoint.r", "plumber.r") %in% appFilesBase),
+    tensorflow = length(c(
+      Sys.glob(file.path(appDir, "*", "saved_model.pb*")),
+      Sys.glob(file.path(appDir, "saved_model.pb*"))
+    )) > 0
+  )
+
+  if (any(satisfiedLayouts)) {
+    return()
+  }
+
+  cli::cli_abort(c(
+    "Cancelling deployment: invalid project layout.",
+    i = "The project should have one of the following layouts:",
+    " " = "1. 'server.R' and 'ui.R' in the application base directory.",
+    " " = "2. 'server.R' and 'www/index.html' in the application base directory.",
```

Comment:
Presenting this list of layouts is misleading, since R Markdown using `runtime: shinyrmd` can have `server.R` and `ui.Rmd` (other Rmd names also fine). The description of layouts also does not hint at R Markdown sites (bookdown).

The `satisfiedLayouts` checks check that the project looks to have the minimal structure for something that _might_ successfully deploy. Listing them in text makes the reader feel like they need to confirm to that exact structure.

I'm aware that you're just moving code around.